agent/internal/taskoutput/test_log.go

```diff
@@ -0,0 +1,299 @@
+	h.createSender = func(ctx context.Context, logPath string) (send.Sender, error) {
+		return output.NewSender(ctx, taskOpts, taskoutput.EvergreenSenderOptions{
+			Local: logger.Task().GetSender(),
+			Parse: h.spec.getParser(),
+		}, logPath)
+	}
+
+	return h
+}
+
+func (h *testLogDirectoryHandler) start(ctx context.Context, dir string) error {
+	h.dir = dir
+	h.watcher = watcher.New()
+	h.watcher.FilterOps(watcher.Create)
+	if err := h.watcher.AddRecursive(h.dir); err != nil {
+		return errors.Wrap(err, "configuring test log directory watcher as recursive")
+	}
+	if err := h.watcher.Ignore(filepath.Join(h.dir, testLogSpecFilename)); err != nil {
+		return errors.Wrap(err, "configuring test log directory watcher to ignore the test log spec file")
+	}
+
+	started := make(chan struct{})
+	go func() {
+		h.watcher.Wait()
+		close(started)
+	}()
+	startErr := make(chan error)
+	go func() {
+		startErr <- h.watcher.Start(time.Millisecond)
+		close(startErr)
+	}()
+
+	select {
+	case <-started:
+	case err := <-startErr:
+		return errors.Wrap(err, "starting test log directory watcher")
+	}
+
+	go h.watch(ctx)
+
+	return nil
+}
+
+func (h *testLogDirectoryHandler) watch(ctx context.Context) {
+	defer func() {
+		h.logger.Execution().Critical(recovery.HandlePanicWithError(recover(), nil, "test log directory watcher"))
+	}()
+
+	workerCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for {
+		select {
+		case event := <-h.watcher.Event:
+			if event.IsDir() {
+				continue
+			}
+
+			h.followFileCount++
+			h.wg.Add(1)
+			go h.followFile(workerCtx, event)
+		case err := <-h.watcher.Error:
+			h.logger.Execution().Critical(errors.Wrap(err, "watching test log directory"))
+			return
+		case <-ctx.Done():
+			h.logger.Execution().Warning("context canceled, exiting test log directory watcher")
+			return
+		case <-h.watcher.Closed:
+			return
+		}
+	}
+}
+
+func (h *testLogDirectoryHandler) getSpecFile() {
+	h.logger.Task().Infof("detected first test log file, getting test log spec file")
+
+	data, err := os.ReadFile(filepath.Join(h.dir, testLogSpecFilename))
+	if err != nil {
+		h.logger.Task().Warning(errors.Wrap(err, "reading test log spec; falling back to default spec"))
+		return
+	}
+	if err = yaml.Unmarshal(data, &h.spec); err != nil {
+		h.logger.Task().Warning(errors.Wrap(err, "unmarshalling test log spec; falling back to default spec"))
+		return
+	}
+
+	if err = h.spec.Format.validate(); err != nil {
+		h.logger.Task().Warning(errors.Wrapf(err, "invalid test log format specified; falling back to default text format"))
+	}
+}
+
+func (h *testLogDirectoryHandler) followFile(ctx context.Context, event watcher.Event) {
+	defer func() {
+		h.logger.Task().Critical(recovery.HandlePanicWithError(recover(), nil, "test log file follower"))
+	}()
+	defer h.wg.Done()
+
+	h.once.Do(h.getSpecFile)
+
+	h.logger.Task().Infof("new test log file '%s' found, initiating automated ingestion", event.Path)
+
+	logPath, err := filepath.Rel(h.dir, event.Path)
```

Comment:
what is this for?

Reply:
We do not want to include the reserved directory name in the name (path in S3) that the log will be stored under. I will add a comment.

Reply:
Ah I get it now.
It took me forever to even understand what Rel _does_